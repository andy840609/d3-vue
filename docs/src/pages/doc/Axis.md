# Axis

Axis are usually added to one of the margin slots in `Cartesian`, which in turn provies several required props to the axis.

```html
<template>
  <d3-cartesian width="900" height="100"
    :x="{type: 'Time', domain: [new Date(2015, 0, 1), new Date(2018, 0, 1)]}"
    :y="{type: 'Linear', domain: [0, 100]}">
    <d3-axis slot="south" orientation="Bottom" slot-scope="props" v-bind="props"/>
    <d3-axis slot="west" orientation="Left" :options="{ticks: {count: 3}}" slot-scope="props" v-bind="props"/>
  </d3-cartesian>
</template>
<!-- axis-demo.vue -->
```

## Props

Name             | Type       | Default      | Required | Description
---------------- | ---------- | ------------ | -------- | -----------------------
`scale`          | Function   |              | true     | D3 scale object, usually set by its parent `Cartesian` component
`orientation`    | String     | 'Bottom'     | true     | Orientation of the ticks. Supported values: 'Bottom', 'Left', 'Right', 'Top'
`title`          | String     |              |          | Title of the axis
`width`          | Number     |              |          | The width of the axis painting area, usually set by its parent `Cartesian` component
`height`         | Number     |              |          | The height of the axis painting area, usually set by its parent `Cartesian` component
`options`        | Object     |              |          | Extra options for the axis

## Slots

None

## Events

None

## Methods

None
