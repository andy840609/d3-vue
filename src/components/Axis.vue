<template>
  <g class="axis"></g>
</template>

<script>
import * as d3 from 'd3'
import Text from './Text.vue'

export default {
  components: { 'd3-text': Text },
  props: {
    title: {
      type: String,
      required: false
    },
    orientation: {
      type: String,
      default: 'Bottom'
    },
    scale: {
      type: Function,
      required: true
    },
    width: {
      type: Number,
      required: true
    },
    height: {
      type: Number,
      required: true
    },
    options: {
      type: Object,
      default: () => {}
    }
  },
  computed: {
    axis () {
      const { orientation, scale, options } = this
      const axis = d3[`axis${orientation}`](scale)
      if (options) {
        const { ticks, tickSize, tickPadding } = options
        if (ticks) axis.ticks(ticks.count, ticks.specifier)
        if (tickSize) axis.tickSize(tickSize)
        if (tickPadding) axis.tickPadding(tickPadding)
      }
      return axis
    }
  },
  watch: {
    axis (val) {
      const { title, orientation, axis, options } = this
      d3.select(this.$el).transition().call(axis)
    },
    title (val) {
      this.update()
    },
    options (val) {
      this.update()
    }
  },
  methods: {
    update () {
      const { title, orientation, axis, options } = this
      const g = d3.select(this.$el).call(axis)

      if (title) {
        switch (orientation) {
          case 'Left':
            if (options && options.titleLastTick) {
              g.select('.tick:last-of-type text')
                .select(function () { return this.parentNode.appendChild(this.cloneNode()) })
                .attr('class', 'label')
                .attr('x', 3)
                .attr('text-anchor', 'start')
                .text(title)
            } else {
              g.append('text')
                .attr('class', 'label')
                .attr('transform', 'rotate(-90)')
                .attr('fill', 'black')
                .attr('y', 6)
                .attr('dy', '.71em')
                .style('text-anchor', 'end')
                .text(title)
            }
            break
          case 'Bottom':
            if (options && options.titleLastTick) {
              g.select('.tick:last-of-type text')
                .select(function () { return this.parentNode.appendChild(this.cloneNode()) })
                .attr('class', 'label')
                .attr('fill', 'black')
                .attr('y', -3)
                .attr('dy', null)
                .text(title)
            } else {
              g.append('text')
                .attr('class', 'label')
                .attr('fill', 'black')
                .attr('x', this.width)
                .attr('y', -6)
                .style('text-anchor', 'end')
                .text(title)
            }
            break
        }
      }
    }
  },
  mounted () {
    this.update()
  }
}
</script>
