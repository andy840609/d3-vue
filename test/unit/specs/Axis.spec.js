import { mount } from '@vue/test-utils'
import * as d3 from 'd3'
import Axis from '@/components/Axis.vue'

const createCmp = propsData => mount(Axis, { propsData })

describe('Axis.vue', () => {
  let wrapper

  it('has the expected html structure', () => {
    wrapper = createCmp({
      width: 600,
      height: 400,
      scaleX: d3.scaleLinear(),
      scaleY: d3.scaleLinear()
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
