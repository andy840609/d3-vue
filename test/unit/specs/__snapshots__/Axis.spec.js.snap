// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Axis.vue has the expected html structure 1`] = `
<g
  class="axis"
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,6V0.5H1.5V6"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(0.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.0
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(0.6,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.1
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(0.7,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.2
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(0.8,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.3
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(0.9,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.4
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(1,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.5
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(1.1,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.6
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(1.2,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.7
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(1.3,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.8
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(1.4,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      0.9
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(1.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="#000"
      y="9"
    >
      1.0
    </text>
  </g>
</g>
`;
